pkg/doc/did/doc.go

```diff
@@ -212,6 +212,16 @@ type Proof struct {
 	Nonce          string     `json:"nonce,omitempty"`
 }
 
+//DID returns the new did
+func (doc *Doc) DID() (string, error) {
```

Comment:
@talwinder50 should be refactored - not all DID documents calculate their DID the same way.

Reply:
then DID() func needs to have a peer document  as an input from somewhere (as we are using public key and authentication (Which are must to have) for creating did.

Reply:
Yes, in the case of peer DIDs the `DID()` function needs as input the DID document itself. That is not the case for other DID methods.

This suggests to me that `did.Doc.ID` should be a function, not a string:

```go
type Doc struct {
	Context        []string
	ID             func() string       /// function
	PublicKey      []PublicKey
	Service        []Service
	Authentication []VerificationMethod
	Created        *time.Time
	Updated        *time.Time
	Proof          *Proof
}
```

Reply:
As per our discussion, when reading from storage the genesis document can be from any DID method (peer DID, sidtree, etc). But we dont know what is the differentiating factor to  compute the specific method DID. The current implementation ties the DOC struct to create the PEER dids only. I have added TODO that we do need to think about how this can be generalized for all other did methods as well.

Reply:
@talwinder50 Generating a DID is a one time task based on Genesis Version of the Document. Not sure it should be fit into a member function of Doc.

Reply:
@rolsonquadras .. I added into a member function of doc. Based on the below comment. 

 @talwinder50 it would be better if the DID is an attribute of the did.Doc. As in, did := doc.DID() where doc is a peer DID document

Basically, all DID documents and their attributes should be under a single unified model

Reply:
I assume you are referring to [this](https://github.com/hyperledger/aries-framework-go/pull/56#discussion_r309697639) comment. But I personally feel one time DID calculation should't part of a member function. It looks like the code here calculates DID dynamically whenever doc.DID is invoked. Documents can be updated the consumers and this will result in different DID on subsequent calls.

Reply:
yea @rolsonquadras i agree. It should be one time thing

Reply:
@llorllale can you please share your insights as well, if there was something else in your mind when you suggested that

Reply:
@rolsonquadras @talwinder50 you're correct - it's not required for `did.Doc` to have an accessor function for its DID. I am naturally inclined towards accessor functions.

To me this seems like a classic case of the Strategy pattern, but I will wait and see how much overlap there is between a parser in package `peer` and other packages.

Reply:
But to be clear - I am strongly opposed to the original design: `did := GetPeerDid(&doc)` - there's no reason why `did.Doc` itself cannot provide the DID. The DID is part of its model.

Reply:
@llorllale The [original version](https://github.com/hyperledger/aries-framework-go/compare/6800c15cc516bf920a81ce2475ab53ee48881a50..f68ebb4b0edf73dec6ac74a4292e3a4f3822206c#diff-32a7b3a4704464a13c867cdbf299342eL22) had ```GeneratePeerDid(doc *did.Doc)```. It was Generate DID and not Get DID(as in, fetch existing DID). You still have reservation on this design?

Reply:
Yes - my objection was not on the name ("GeneratePeerDid") - but the fact that `doc.ID` didn't just give me the DID. And since `GeneratePeerDid` is exported, I assumed that the design intended for *users* to call `GeneratePeerDid(doc)` - which is something I strongly oppose